### ISSUE TITLE
<Form> submitter is serialized out of tree order

### ISSUE THREAD

**Issue by jenseng** [remix-run/remix#4342]
### What version of Remix are you using?

1.7.2

### Steps to Reproduce

1. Make a `<Form>` like so:

```jsx
<Form>
  <button type="submit" name="foo" value="override">Override</button>
  <input type="hidden" name="foo" value="default" />
</Form>
```

2. Click the "Override" button

### Expected Behavior

The form fields should be serialized in tree order, i.e. `foo=override&foo=default` (as is the case with a vanilla `<form>`).

See [Constructing the form data set](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-form-data-set) in the HTML spec. Submitted fields should be sent in the order they appear in the DOM, including the `submitter` (i.e. the submit button that was clicked).

### Actual Behavior

The form is serialized as `foo=default&foo=override`.

This is problematic on the server-side if you are relying on the order to indicate precedence. `formData.get("foo")` will yield different values, depending on whether the form was submitted by Remix or native browser behavior.

See also https://github.com/remix-run/remix/pull/3611; prior to that fix it was broken in a different way (it would only serialize `foo=default`)

**Comment by jenseng** [comment]
On a related note, `<Form>` serializes image submit buttons in the wrong order **and** with the wrong key/values.

For example, given: `<input type="image" name="image" src="..." />`. If you click it, it gets serialized as `image=` at the very end, rather than the expected `image.x=X&image.y=Y` in tree order. See [Step 3.3](https://www.w3.org/TR/2014/REC-html5-20141028/forms.html#constructing-form-data-set).

Probably worth lumping in to this bug report, as the root cause is the same, and the fix should be too 🤞

**Comment by jenseng** [comment]
I found some more differences between how `<Form>` and `<form>` behave:

1. In `<Form>` submissions, Safari includes the submit button value twice (once in tree order, once at the end) due to [this bug](https://bugs.webkit.org/show_bug.cgi?id=239070)
2. If you have file inputs and the `encType` is `application/x-www-form-urlencoded` (explicit or implicit):
   * `<form>` will send each file entry's `name` as the `value`, as per [the spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#convert-to-a-list-of-name-value-pairs)
   * `<Form>` will throw an error `File inputs are not supported with encType "application/x-www-form-urlencoded" ...`; if this `invariant` check is removed, no `value` gets submitted.

**Comment by machour** [comment]
Great stuff @jenseng, thanks!

**Comment by jenseng** [comment]
As a workaround for non-image submitters, you can use a component like so that does some hidden input hijinks:

```tsx
function Button(props: DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>) {
  const hiddenRef = useRef<HTMLInputElement>(null);
  const buttonRef = useRef<HTMLButtonElement>(null);
  if ((props.type ?? "submit") === "submit" && props.name) {
    return (
      <>
        <input ref={hiddenRef} type="hidden" name={props.name} value={props.value} disabled />
        <button
          ref={buttonRef}
          {...props}
          onClick={(e) => {
            hiddenRef.current!.disabled = false;
            buttonRef.current!.name = "";
            setTimeout(() => {
              hiddenRef.current!.disabled = true;
              buttonRef.current!.name = props.name!;
            });
            props.onClick?.(e);
          }}
        />
      </>
    );
  } else {
    return <button {...props} />;
  }
}
```

Since the hijinks are done in the `onClick` handler, the button still has the right attributes to work before/without JavaScript

**Comment by brophdawg11** [comment]
Thank you for the detailed use cases and spec references @jenseng - these are incredibly helpful 🙌

**Comment by brophdawg11** [comment]
I think this should be resolved when we release Remix `1.18.0` now that the relevant PRs are merged in RR.  The latest `FormData` submitter PR should be released in RR `6.14.0`
